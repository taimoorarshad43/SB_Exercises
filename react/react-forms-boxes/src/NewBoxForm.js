import React, { useState } from "react";
import { v4 as uuid } from 'uuid';

function NewBoxForm({ createBox }) {
  const [formData, setFormData] = useState({
    height: "",
    width: "",
    backgroundColor: ""
  });

  const handleChange = evt => {
    const { name, value } = evt.target;
    setFormData(formData => ({
      ...formData,
      [name]: value
    }));
  };

  const gatherInput = evt => {
    evt.preventDefault();
    createBox({ ...formData, id: uuid() });
    setFormData({ height: "", width: "", backgroundColor: "" });
  };

  return (
    <div style={{ marginBottom: '20px', padding: '20px', border: '1px solid #ccc', borderRadius: '5px' }}>
      <h3>Add a New Box</h3>
      <form onSubmit={gatherInput}>
        <div style={{ marginBottom: '10px' }}>
          <label htmlFor="height">Height: </label>
          <input
            onChange={handleChange}
            type="text"
            name="height"
            value={formData.height}
            id="height"
            placeholder="e.g., 5"
            style={{ marginLeft: '5px', padding: '5px' }}
          />
        </div>
        <div style={{ marginBottom: '10px' }}>
          <label htmlFor="width">Width: </label>
          <input
            onChange={handleChange}
            type="text"
            name="width"
            id="width"
            value={formData.width}
            placeholder="e.g., 5"
            style={{ marginLeft: '5px', padding: '5px' }}
          />
        </div>
        <div style={{ marginBottom: '10px' }}>
          <label htmlFor="backgroundColor">Background Color: </label>
          <input
            onChange={handleChange}
            type="text"
            name="backgroundColor"
            value={formData.backgroundColor}
            id="backgroundColor"
            placeholder="e.g., red"
            style={{ marginLeft: '5px', padding: '5px' }}
          />
        </div>
        <button type="submit" style={{ padding: '8px 16px', backgroundColor: '#007bff', color: 'white', border: 'none', borderRadius: '3px', cursor: 'pointer' }}>
          Add a new box!
        </button>
      </form>
    </div>
  );
}

export default NewBoxForm;
