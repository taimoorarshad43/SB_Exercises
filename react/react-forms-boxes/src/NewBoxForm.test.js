import React from "react";
import { render, fireEvent } from "@testing-library/react";
import NewBoxForm from "./NewBoxForm";

it("renders without crashing", function() {
  render(<NewBoxForm />);
});

it("matches snapshot", function() {
  const { asFragment } = render(<NewBoxForm />);
  expect(asFragment()).toMatchSnapshot();
});

it("updates form data when inputs change", function() {
  const { getByLabelText } = render(<NewBoxForm />);
  
  const heightInput = getByLabelText("Height:");
  const widthInput = getByLabelText("Width:");
  const colorInput = getByLabelText("Background Color:");
  
  fireEvent.change(heightInput, { target: { value: "10" } });
  fireEvent.change(widthInput, { target: { value: "15" } });
  fireEvent.change(colorInput, { target: { value: "green" } });
  
  expect(heightInput.value).toBe("10");
  expect(widthInput.value).toBe("15");
  expect(colorInput.value).toBe("green");
});

it("calls createBox with form data on submit", function() {
  const createBoxMock = jest.fn();
  const { getByLabelText, getByText } = render(<NewBoxForm createBox={createBoxMock} />);
  
  const heightInput = getByLabelText("Height:");
  const widthInput = getByLabelText("Width:");
  const colorInput = getByLabelText("Background Color:");
  const submitButton = getByText("Add a new box!");
  
  fireEvent.change(heightInput, { target: { value: "10" } });
  fireEvent.change(widthInput, { target: { value: "15" } });
  fireEvent.change(colorInput, { target: { value: "green" } });
  fireEvent.click(submitButton);
  
  expect(createBoxMock).toHaveBeenCalledWith({
    height: "10",
    width: "15",
    backgroundColor: "green",
    id: expect.any(String)
  });
});

it("clears form after submission", function() {
  const createBoxMock = jest.fn();
  const { getByLabelText, getByText } = render(<NewBoxForm createBox={createBoxMock} />);
  
  const heightInput = getByLabelText("Height:");
  const widthInput = getByLabelText("Width:");
  const colorInput = getByLabelText("Background Color:");
  const submitButton = getByText("Add a new box!");
  
  fireEvent.change(heightInput, { target: { value: "10" } });
  fireEvent.change(widthInput, { target: { value: "15" } });
  fireEvent.change(colorInput, { target: { value: "green" } });
  fireEvent.click(submitButton);
  
  expect(heightInput.value).toBe("");
  expect(widthInput.value).toBe("");
  expect(colorInput.value).toBe("");
});
