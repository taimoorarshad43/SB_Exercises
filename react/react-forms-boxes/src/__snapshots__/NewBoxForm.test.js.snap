// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    style="margin-bottom: 20px; padding: 20px; border: 1px solid #ccc; border-radius: 5px;"
  >
    <h3>
      Add a New Box
    </h3>
    <form>
      <div
        style="margin-bottom: 10px;"
      >
        <label
          for="height"
        >
          Height: 
        </label>
        <input
          id="height"
          name="height"
          placeholder="e.g., 5"
          style="margin-left: 5px; padding: 5px;"
          type="text"
          value=""
        />
      </div>
      <div
        style="margin-bottom: 10px;"
      >
        <label
          for="width"
        >
          Width: 
        </label>
        <input
          id="width"
          name="width"
          placeholder="e.g., 5"
          style="margin-left: 5px; padding: 5px;"
          type="text"
          value=""
        />
      </div>
      <div
        style="margin-bottom: 10px;"
      >
        <label
          for="backgroundColor"
        >
          Background Color: 
        </label>
        <input
          id="backgroundColor"
          name="backgroundColor"
          placeholder="e.g., red"
          style="margin-left: 5px; padding: 5px;"
          type="text"
          value=""
        />
      </div>
      <button
        style="padding: 8px 16px; background-color: rgb(0, 123, 255); color: white; border-radius: 3px; cursor: pointer;"
        type="submit"
      >
        Add a new box!
      </button>
    </form>
  </div>
</DocumentFragment>
`;
